[clang][CodeGen] Zero init unspecified fields in initializers in C

yabinc opened the pull request:
When an initializer is provided to a variable, the Linux kernel relied
on the compiler to zero-initialize unspecified fields, as clarified in
https://www.spinics.net/lists/netdev/msg1007244.html.

But clang doesn't guarantee this:
1. For a union type, if an empty initializer is given, clang only
   initializes bytes for the first field, left bytes for other (larger)
   fields are marked as undef. Accessing those undef bytes can lead
   to undefined behaviors.
2. For a union type, if an initializer explicitly sets a field, left
   bytes for other (larger) fields are marked as undef.
3. When an initializer is given, clang doesn't zero initialize padding.

So this patch makes the following change:
1. In C, when an initializer is provided for a variable, zero-initialize
   undef and padding fields in the initializer.
2. Document the change in LanguageExtensions.rst.

As suggested in https://github.com/llvm/llvm-project/issues/78034#issuecomment-2183437928,
the change isn't required by C23, but it's standards conforming to do
so.

Fixes: https://github.com/llvm/llvm-project/issues/97459

efriedma-quic commented:
The only place the standard requires padding to be zero-initialized is in "default initialization", which only applies to members which don't have an initializer.  So, for example, if you have `struct X { int x; char y; } z = {1};`, none of the padding is required to be initialized; the only default initialization is of "y", which doesn't have any padding.  Similarly, if you have  `union X { char x; int y; } z = {1};`, none of the padding is required to be initialized; there's no default initialization at all.  On the other hand, if you write `union X { char x; int y; } z = {};`, that padding is required to be initialized.

There's probably some argument for doing what this patch does, in addition to what the standard requires.  But to avoid any confusion, please file a separate bug for that, and make sure the comments in the patch distinguish between what the standard requires and what we actually implement.

yabinc commented:
@efriedma-quic, Thanks for the suggestion!
I have created a new bug in https://github.com/llvm/llvm-project/issues/97459, and also updated commit message to clarify that this is needed by the linux kernel instead of c23 standard.

yabinc commented:
@efriedma-quic, @rjmccall , @AaronBallman , @nickdesaulniers , ping for review or comments?

AaronBallman reviewed (commented):
Thank you for working on this! I'll leave the codegen code review to John and Eli, but I would appreciate explicit documentation in `clang/docs/LanguageExtensions.rst` so that we explicitly document this behavior and users (including the Linux kernel) can rely on it longterm.

AaronBallman commented:
Ping @rjmccall @efriedma-quic -- I'd like to see this get into the release so that the Linux kernel can be built with Clang 19.

pirama-arumuga-nainar commented:
Ping @rjmccall @efriedma-quic.  Any comments on this patch?

yabinc commented:
Test case showing difference in behavior w.r.t. gcc (without this patch) https://godbolt.org/z/9fPe78cEz

rjmccall commented on clang/lib/CodeGen/CGDecl.cpp:
```diff
@@ -361,6 +368,13 @@ CodeGenFunction::AddInitializerToStaticVarDecl(const VarDecl &D,
+  if (!getLangOpts().CPlusPlus) {
+    // In C, when an initializer is given, the Linux kernel relies on clang to
+    // zero-initialize all members not explicitly initialized, including padding
+    // bits.
```
Comments in the compiler generally shouldn't refer to specific projects.  We are making a guarantee based on our interpretation of the standard.

Similarly, the PR title is also inappropriate.  It is fine to note in the PR description that Linux relies on this for correctness, but the PR title should describe the semantic change.

rjmccall replied:
Also, I think we should do this unconditionally in all language modes.  It's not like this is forbidden in C++.

yabinc replied:
> But I assume that similar code is required in order to explicitly zero out any padding when dynamically initializing a local variable.

I guess you mean variables assigned by Compound literals after variable definition. An example is in https://godbolt.org/z/caaEKnjWe. From the ast dump, it is treated as an assignment from InitListExpr. The source code seems to be in AggExprEmitter::VisitCXXParenListOrInitListExpr() in https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGExprAgg.cpp#L1757. Each field of the InitListExpr is stored to the variable separately. One solution is to always do memset in CheckAggExprForMemSetUse() in https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGExprAgg.cpp#L1974. The memset can be optimized away later if not needed. 
Another place is compound literals for global values. It seems handled in CGM.GetAddrOfConstantCompoundLiteral().
To not introduce a big change at once, Shall I handle them in a follow up patch?

> Also, is adding constant padding fields retroactively really the best way to handle this instead of just expecting the constant emitter to fill them in itself?

Currently I limit the change to C and reuse code for trivial auto var zero init. So it's easier to do it outside the constant emitter. If the change can be extended to C++, I feel it's better to always do it in the constant emitter. 

> Comments in the compiler generally shouldn't refer to specific projects. We are making a guarantee based on our interpretation of the standard.

The standard is ambiguous on this, so can't be used as a reason for the change. Please let me know if you have better idea about how to reword the comment.

> Similarly, the PR title is also inappropriate. It is fine to note in the PR description that Linux relies on this for correctness, but the PR title should describe the semantic change.

Changed the PR title back to the commit title, which describes the semantic change.

> Also, I think we should do this unconditionally in all language modes. It's not like this is forbidden in C++.

I agree. But when I was trying to do it before, I need to fix a lot more tests. And the Linux kernel only has C code. I don't have strong reason to push this to C++. So I want to limit the change to C at first. If reviewers think we should also do it in C++, shall I do it in a follow up change?

rjmccall replied:
> > But I assume that similar code is required in order to explicitly zero out any padding when dynamically initializing a local variable.
> 
> I guess you mean variables assigned by Compound literals after variable definition. An example is in https://godbolt.org/z/caaEKnjWe. 

No, I mean initializing a local aggregate with a non-constant initializer, e.g.

```
void g(char c) {
  struct A a = { c, 2 };
}
```

Do we already zero the padding correctly in this case?

yabinc replied:
> Do we already zero the padding correctly in this case?

No, the non-constant initializer goes through the same implemented path as "variables assigned by Compound literals after variable definition". I guess I'd better do it in the same patch. Do you have any suggestions on how to implement it? Otherwise I'll try always do memset in CheckAggExprForMemSetUse() in https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGExprAgg.cpp#L1974.

rjmccall replied:
> The standard is ambiguous on this, so can't be used as a reason for the change. Please let me know if you have better idea about how to reword the comment.

I suggest quoting the standard, identifying the ambiguity, and then saying that we've chosen to interpret it as requiring all padding in the object to be zeroed.  Since there are multiple paths in the compiler that need this logic, you may want to write this in a single place and then cross-reference it from other places.

> I agree. But when I was trying to do it before, I need to fix a lot more tests. And the Linux kernel only has C code. I don't have strong reason to push this to C++. So I want to limit the change to C at first. If reviewers think we should also do it in C++, shall I do it in a follow up change?

I don't mind staging it in over multiple commits in principle.  If you'd like to switch to an implementation that adds explicit padding fields when building the original constant, you *could* still stage that and only add those fields in certain language modes.

> No, the non-constant initializer goes through the same implemented path as "variables assigned by Compound literals after variable definition". I guess I'd better do it in the same patch. Do you have any suggestions on how to implement it?

We probably do need to consider it when deciding whether to memset, but naively I'd expect that you'd just emit little memsets for each individual piece of padding when we recognize that there's a gap between successive fields.  Does that not work?

yabinc replied:
> I suggest quoting the standard, identifying the ambiguity, and then saying that we've chosen to interpret it as requiring all padding in the object to be zeroed. Since there are multiple paths in the compiler that need this logic, you may want to write this in a single place and then cross-reference it from other places.

Done.

> I don't mind staging it in over multiple commits in principle. If you'd like to switch to an implementation that adds explicit padding fields when building the original constant, you could still stage that and only add those fields in certain language modes.

I am OOO next week, will try it in ConstantEmitter after that.

> We probably do need to consider it when deciding whether to memset, but naively I'd expect that you'd just emit little memsets for each individual piece of padding when we recognize that there's a gap between successive fields. Does that not work?

It works. Done.

yabinc replied:
> I don't mind staging it in over multiple commits in principle. If you'd like to switch to an implementation that adds explicit padding fields when building the original constant, you could still stage that and only add those fields in certain language modes.

Changed as suggested, adding explicit padding fields when building the original constant.

efriedma-quic replied:
Adding handling for C++ is tricky: C++ allows fields and base classes to overlap.  For static initialization, it's straightforward to check for overlap, but for dynamic initialization it's problematic.  In some cases (AggValueSlot::DoesNotOverlap), we can prove there is no overlap, but in general it's ambiguous.   In such cases, we have to leave padding uninitialized.  So I think sticking to C for the first iteration makes sense; we'll have to carefully consider exactly what guarantees we can provide for C++.

rjmccall reviewed (commented):
Is this really all that's required?  It looks like you're just filling in explicit zero padding when emitting constant initializers.  That should steer clear of any possibility that LLVM would treat the padding as `undef` for optimization purposes (surely `undef` bytes are still emitted as zero when rendering the initializer in assembly? it doesn't hurt to be more explicit, of course), so okay, it's good to do.  But I assume that similar code is required in order to explicitly zero out any padding when dynamically initializing a local variable.

Also, is adding constant padding fields retroactively really the best way to handle this instead of just expecting the constant emitter to fill them in itself?

yabinc commented:
Sorry for using force push to resolve merge conflicts. The patch has been updated as suggested, and is ready for review.

yabinc commented:
@rjmccall and @efriedma-quic, ping for review?

pirama-arumuga-nainar commented:
@rjmccall and @efriedma-quic, ping for review?

@AaronBallman suggested this to be included in LLVM 19, which is slated for final release next week.  Is the window for backports still open?

AaronBallman commented:
> @rjmccall and @efriedma-quic, ping for review?
> 
> @AaronBallman suggested this to be included in LLVM 19, which is slated for final release next week. Is the window for backports still open?

No, unfortunately I think it's too late to make 19.1 -- release managers are only taking extremely safe and important fixes, and while this is extremely important, it's a pretty sizable change that needs a good amount of testing. However, if we do a 19.2, this could perhaps qualify.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1745,13 +1748,48 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+        Address BaseLoc = Dest.getAddress().withElementType(CGF.Int8Ty);
+        Address LeftLoc =
+            CGF.Builder.CreateConstGEP(BaseLoc, LeftSize.getQuantity());
+        CGF.Builder.CreateMemSet(LeftLoc, CGF.Builder.getInt8(0), LeftSizeVal,
```
If we don't have an initializer, instead of explicitly init'ing the padding, we can just `EmitNullInitializationToLValue(DestLV)`.

yabinc replied:
Done.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1745,13 +1748,48 @@ void AggExprEmitter::VisitCXXParenListOrInitListExpr(
+            CGF.Builder.getInt64((Offset - SizeSoFar).getQuantity());
+        Address PaddingLoc =
+            CGF.Builder.CreateConstGEP(BaseLoc, SizeSoFar.getQuantity());
+        CGF.Builder.CreateMemSet(PaddingLoc, CGF.Builder.getInt8(0),
```
Use a lambda to unify the code a bit?

yabinc replied:
Done.

efriedma-quic commented on clang/test/CodeGen/linux-kernel-struct-union-initializer.c:
```diff
@@ -0,0 +1,219 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-unknown -std=gnu11 -verify -emit-llvm %s -o - | FileCheck %s
```
update_cc_test_checks.py, maybe?  Unless it doesn't generate clear checks for some reason.

yabinc replied:
I didn't know that. Now use the script to generate checks.

yabinc commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -5819,3 +5819,24 @@ specify the starting offset to begin embedding from. The resources is treated
+Union and aggregate initialization in C
+=======================================
+
+In C23 (N2900), when an object is initialized from initializer ``= {}``, all
```
Added.

yabinc commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -42,6 +42,66 @@ using namespace CodeGen;
+    if (PadSize == CharUnits::One())
+      return llvm::ConstantInt::get(CGM.CharTy, 0);
+    llvm::Type *Ty = llvm::ArrayType::get(CGM.CharTy, PadSize.getQuantity());
+    return llvm::ConstantAggregateZero::get(Ty);
```
Done.

yabinc commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -715,6 +772,11 @@ bool ConstStructBuilder::Build(const InitListExpr *ILE, bool AllowOverwrite) {
       return false;
 
+  // See comment in getPadding().
+  bool ZeroInitPadding = !CGM.getLangOpts().CPlusPlus;
```
Done.

efriedma-quic reviewed (commented):
It probably makes sense to add a helper like the following to CodeGenModule or something like that:

```
bool shouldZeroInitPadding() { 
  // Comment explaining reasoning for this rule
  return !CGM.getLangOpts().CPlusPlus;
}
```

So we don't have `// See comment in getPadding().` all over the place.

----

Is there any relevant code using EmitNullConstant()?  As far as I can tell, this patch doesn't  zero-fill implicit padding in null constants (like in `struct { int x; long y; }`.

yabinc commented:
> It probably makes sense to add a helper like the following to CodeGenModule or something like that:
> 
> ```
> bool shouldZeroInitPadding() { 
>   // Comment explaining reasoning for this rule
>   return !CGM.getLangOpts().CPlusPlus;
> }
> ```
> 
> So we don't have `// See comment in getPadding().` all over the place.

Done.

> 
> Is there any relevant code using EmitNullConstant()? As far as I can tell, this patch doesn't zero-fill implicit padding in null constants (like in `struct { int x; long y; }`.

I ran all tests in clang/tests and found two places using it:
1. One is used in CodeGenFunction::EmitNullInitialization(Address DestPtr, QualType Ty), when !CGM.getTypes().isZeroInitializable(Ty). This seems to only happen in C++ code (when a class contains a pointer pointing
  to data member).
2. Another is used in CodeGenModule::getOrCreateStaticVarDecl(), which is further called in three places:
  a. CodeGenFunction::AddInitializerToStaticVarDecl(): it later calls ConstantEmitter::tryEmitForInitializer() if an initializer is used.
  b. CodeGenFunction::EmitDeclRefLValue(const DeclRefExpr *E) in CGExpr.cpp 
  c. ConstantLValueEmitter::tryEmitBase(const APValue::LValueBase &base) in CGExprConstant.cpp
  Not sure what b and c are doing.
When running tests for non c++ code, I only found the 2.a path, used in CodeGenOpenCL/amdgpu-nullptr.cl.
So I didn't find any C code using EmitNullConstant(), that results in uninitialized fields with initializers.

For the implementation of EmitNullConstant(), it returns ConstantStruct::get(llvm::ConstantStruct::get(structure, elements). In ConstantStruct::get(), if all elements are zero, then it returns ConstantAggregateZero::get(ST).
From  https://llvm.org/docs/LangRef.html:
`The string ‘zeroinitializer’ can be used to zero initialize a value to zero of any type, including scalar and [aggregate](https://llvm.org/docs/LangRef.html#t-aggregate) types. This is often used to avoid having to print large zero initializers (e.g. for large arrays) and is always exactly equivalent to using explicit zero initializers.`

From the above words, I feel using `%struct.S2 zeroinitializer` may not guarantee the padding in the struct are zero initialized. But if it is used for a static variable created by CodeGenModule::getOrCreateStaticVarDecl(), it's
as good as a static/global variable without initializer.

yabinc commented:
@rjmccall and @efriedma-quic, ping for review?

pirama-arumuga-nainar commented:
@rjmccall and @efriedma-quic, do you have any other comments on this PR?

efriedma-quic commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -870,6 +904,15 @@ bool ConstStructBuilder::Build(const APValue &Val, const RecordDecl *RD,
+      if (!DoZeroInitPadding(Layout, FieldNo, *Field, SizeSoFar,
+                             IsFlexibleArray, AllowOverwrite))
+        return false;
+      if (IsFlexibleArray)
```
I think I've figured out the way IsFlexibleArray works, but it's sort of awkward.  Do you really need to use a different codepath for flexible-array fields, vs. ordinary fields?  It seems like it would be simpler to just use getTypeAllocSize() in both cases.  And the name "IsFlexibleArray" is a bit confusing because isn't really reliably detecting flexible arrays; there are other zero-size fields.

yabinc replied:
Changed IsFlexibleArray to FieldSize.
If you mean to use getTypeAllocSize() in DoZeroInitPadding(),
I don't know how to replace ASTContext::getTypeSizeInChars(clang::QualType T) with DataLayout::getTypeAllocSize(llvm::Type *Ty).

If you mean to always use CGM.getDataLayout().getTypeAllocSize(EltInit->getType()), there is a case when EltInit isn't avaiable (when emitting a DesignatedInitUpdateExpr with a nested InitListExpr).

efriedma-quic reviewed (approved):
LGTM

I think we'll need to watch out for performance regressions... in most cases, we can probably zero-pad cheaply, but there might be cases where it isn't cheap, or cases where it should be cheap but the LLVM backend can't figure it out.  There's a limit to how much we can do before this lands... but maybe helpful if you can show Linux kernel codesize as a reference, to show there aren't massive issues.

yabinc committed:
Zero init unspecified fields in initializers in C (#97459)

hen an initializer is provided to a variable, the Linux kernel relied
on the compiler to zero-initialize unspecified fields, as clarified in
https://www.spinics.net/lists/netdev/msg1007244.html.

But clang doesn't guarantee this:
1. For a union type, if an empty initializer is given, clang only
   initializes bytes for the first field, left bytes for other (larger)
   fields are marked as undef. Accessing those undef bytes can lead
   to undefined behaviors.
2. For a union type, if an initializer explicitly sets a field, left
   bytes for other (larger) fields are marked as undef.
3. When an initializer is given, clang doesn't zero initialize padding.

So this patch makes the following change:
1. In C, when an initializer is provided for a variable, zero-initialize
   undef and padding fields in the initializer.
2. Document the change in LanguageExtensions.rst.

As suggested in https://github.com/llvm/llvm-project/issues/78034#issuecomment-2183437928,
the change isn't required by C23, but it's standards conforming to do
so.

yabinc commented:
> but maybe helpful if you can show Linux kernel codesize as a reference, to show there aren't massive issues.

I tested it on android-mainline, which closely follows linux upstream. The size change is very small.
.text 0x1139000 -> 0x1138000
.init.text 0x061f08 -> 0x062050
.data 0x20bec8 -> 0x20bf08

The linux kernel enables -ftrivial-auto-var-init=zero by default, which I guess already covers most (if not all) initialization instructions.

yabinc commented:
I manually squashed all the commits. Because there is a merge conflict when I was trying git merge --squash locally.

hubert-reinterpretcast commented:
@AaronBallman, considering the concern of potential performance regressions ("unnecessary" initialization of automatic-storage duration objects), should it be the case that there is no option control?

It was also the case that the prior Clang behaviour (with `-ftrivial-auto-var-init=pattern` [1]: https://godbolt.org/z/x994WMT6h) had some utility for identifying potential portability issues with a program. Option control would make me much happier about this change.

[1]:
A string of `0xaa` is observed using `-ftrivial-auto-var-init=pattern` using Clang 19.1:
```c
typedef union { int x; long long y; } U;

int printf(const char *, ...);
int main(void) {
  U u = { 0 };
  typedef unsigned char *ucp;
  for (ucp p = (ucp)&u; p < (ucp)&1[&u]; ++p) {
    printf("%02x", (unsigned)*p);
  }
  printf("\n");
}
```

efriedma-quic commented:
(For reference, this was reapplied in #110051.)

I'm not anticipating significant performance regressions here in normal usage: all the bits that are getting initialized are padding adjacent to initialized fields, so I expect the additional initialization will usually be free, and cheap in edge cases where it isn't free.  There's a chance of regressions in some edge cases, but that doesn't seem like enough to warrant an option.

Could you clarify which parts of this you'd want to disable?  Handling for all variables?  Handling for local variables?  Handling for variables where C23 rules don't require zero-init?

hubert-reinterpretcast commented:
Handling for local variables (not initialized with `{}`). The performance impact of guaranteeing the extension is artificially reduced because the baseline code generation uses `memset` over the entire union when the size is large, but initializing the bytes after the first member is not mandated.

yabinc commented:
According to the C23 standard, initializing a union variable with empty braces {} only guarantees the initialization of the first member and any padding after all members. This behavior deviates from common intuition and caused a bug when compiling the Linux kernel.
Based on the discussion in https://github.com/llvm/llvm-project/issues/78034, this patch aims to zero-initialize all bytes after the first member.
If this change introduces significant regressions in edge cases, a fallback strategy is to replace brace initialization with explicit assignment statements for the specific fields requiring initialization.

efriedma-quic commented:
> the baseline code generation uses memset over the entire union when the size is large

That looks like an unintended side-effect of CheckAggExprForMemSetUse: it's supposed to check how much of the initialization is non-zero, but it's assuming all padding needs to be zeroed.  Which we weren't explicitly doing before, but I guess it's convenient here. And given nobody has filed a bug report for this (I think?), it doesn't seem this memset is causing significant issues for anyone.

yabinc commented:
I am not sure if it is worth using Store instead of memset for 8 bytes or fewer, as below:
`  if (BitSize <= 64) {
    Builder.CreateStore(Builder.getIntN(BitSize, 0), Addr);
  }`

As shown in https://godbolt.org/z/PEW1zxc1G, memset is not optimized away in -O2 -emit-llvm, but it is optimized to stores in the arm64/x86_64 backend with -O2.

efriedma-quic commented:
There are two possible alternative sequences we could generate there:

```
int test(int x, int y) {
    struct S s;
    __builtin_memset(&s, 0, 8);
    s.x = x;
    s.y = y;
    f(&s);
}
int test2(int x, int y) {
    struct S s;
    long long dummy = (unsigned char)x;
    __builtin_memcpy(&s, &dummy, 8);
    s.x = x;
    s.y = y;
    f(&s);
}
```

Whether one of them is better might depend on the context... for test2, the zero-extend can be folded away in some cases.

I'd prefer to handle this on the LLVM side if possible (memcpyopt, or something like that); it's probably tricky to catch all the relevant cases in clang.

efriedma-quic commented:
(There's basically no benefit to just directly replacing a memset with an equivalent store; most relevant LLVM optimizations are memset-aware.)
